prow/cmd/generic-autobumper/main.go

```diff
@@ -19,47 +19,467 @@ package main
+	UpdateFile(tagPicker func(imageHost, imageName, currentTag string) (string, error), path string, imageFilter *regexp.Regexp) error
+	GetReplacements() map[string]string
+	AddToCache(image, newTag string)
+	TagExists(imageHost, imageName, currentTag string) (bool, error)
+}
+
+func updateReferences(imageBumperCli imageBumper, filterRegexp *regexp.Regexp, o *Options) (map[string]string, error) {
+	var tagPicker func(string, string, string) (string, error)
+	var err error
+
+	switch o.TargetVersion {
+	case latestVersion:
+		tagPicker = imageBumperCli.FindLatestTag
+	case upstreamVersion, upstreamStagingVersion:
+		tagPicker, err = upstreamImageVersionResolver(o, o.TargetVersion, parseUpstreamImageVersion, imageBumperCli)
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve the %s image version: %w", o.TargetVersion, err)
+		}
+	default:
+		tagPicker = func(imageHost, imageName, currentTag string) (string, error) { return o.TargetVersion, nil }
+	}
+
+	updateFile := func(name string) error {
+		logrus.Infof("Updating file %s", name)
+		if err := imageBumperCli.UpdateFile(tagPicker, name, filterRegexp); err != nil {
+			return fmt.Errorf("failed to update the file: %w", err)
+		}
+		return nil
+	}
+	updateYAMLFile := func(name string) error {
+		if strings.HasSuffix(name, ".yaml") && !isUnderPath(name, o.ExcludedConfigPaths) {
+			return updateFile(name)
+		}
+		return nil
+	}
+
+	// Updated all .yaml files under the included config paths but not under excluded config paths.
+	for _, path := range o.IncludedConfigPaths {
+		info, err := os.Stat(path)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get the file info for %q", path)
+		}
+		if info.IsDir() {
+			err := filepath.Walk(path, func(subpath string, info os.FileInfo, err error) error {
+				return updateYAMLFile(subpath)
+			})
+			if err != nil {
+				return nil, fmt.Errorf("failed to update yaml files under %q: %w", path, err)
+			}
+		} else {
+			if err := updateYAMLFile(path); err != nil {
+				return nil, fmt.Errorf("failed to update the yaml file %q: %w", path, err)
+			}
+		}
+	}
+
+	// Update the extra files in any case.
+	for _, file := range o.ExtraFiles {
+		if err := updateFile(file); err != nil {
+			return nil, fmt.Errorf("failed to update the extra file %q: %w", file, err)
+		}
+	}
+
+	return imageBumperCli.GetReplacements(), nil
+}
+
+// used by updateReferences
+func upstreamImageVersionResolver(
+	o *Options, upstreamVersionType string, parse func(upstreamAddress, prefix string) (string, error), imageBumperCli imageBumper) (func(imageHost, imageName, currentTag string) (string, error), error) {
+	upstreamVersions, err := upstreamConfigVersions(upstreamVersionType, o, parse)
+	if err != nil {
+		return nil, err
+	}
+
+	return func(imageHost, imageName, currentTag string) (string, error) {
+		imageFullPath := imageHost + "/" + imageName + ":" + currentTag
+		for prefix, version := range upstreamVersions {
+			if strings.HasPrefix(imageFullPath, prefix) {
```

Comment:
Please try to minimize indenting, so I would rewrite this as:

```golang
if !strings.HasPrefix(imageFullPath, prefix) {
  continue
}
exists, err := imageBumperCli...
...
```

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

Reply:
done